pkg/grpc/universalapi/api_workflow.go

```diff
@@ -50,175 +47,196 @@ func (a *UniversalAPI) GetWorkflowAlpha1(ctx context.Context, in *runtimev1pb.Ge
-func (a *UniversalAPI) workflowActivity(ctx context.Context, in *runtimev1pb.WorkflowActivityRequest, method func(context.Context, *workflows.WorkflowReference) error, methodErr messages.APIError) (*runtimev1pb.WorkflowActivityResponse, error) {
-	if in.InstanceId == "" {
-		err := messages.ErrMissingOrEmptyInstance
+	req := &workflows.PauseRequest{
+		InstanceID: in.InstanceId,
+	}
+	if err := workflowComponent.Pause(ctx, req); err != nil {
+		err = messages.ErrPauseWorkflow.WithFormat(in.InstanceId, err)
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
+	return emptyResponse, nil
+}
 
-	if in.WorkflowComponent == "" {
-		err := messages.ErrNoOrMissingWorkflowComponent
+// ResumeWorkflowAlpha1 is the API handler for resuming a workflow
+func (a *UniversalAPI) ResumeWorkflowAlpha1(ctx context.Context, in *runtimev1pb.ResumeWorkflowRequest) (*emptypb.Empty, error) {
+	emptyResponse := &emptypb.Empty{}
+	if err := a.validateInstanceID(in.InstanceId, false /* isCreate */); err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
 
-	if _, ok := a.CompStore.GetWorkflow(in.WorkflowComponent); !ok {
-		err := messages.ErrWorkflowComponentDoesNotExist.WithFormat(in.WorkflowComponent)
+	workflowComponent, err := a.getComponent(in.WorkflowComponent)
+	if err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
 
-	req := workflows.WorkflowReference{
+	req := &workflows.ResumeRequest{
 		InstanceID: in.InstanceId,
 	}
+	if err := workflowComponent.Resume(ctx, req); err != nil {
+		err = messages.ErrResumeWorkflow.WithFormat(in.InstanceId, err)
+		a.Logger.Debug(err)
+		return emptyResponse, err
+	}
+	return emptyResponse, nil
+}
 
-	err := method(ctx, &req)
-	if err != nil {
-		err = methodErr.WithFormat(in.InstanceId)
+func (a *UniversalAPI) validateInstanceID(instanceID string, isCreate bool) error {
+	if instanceID == "" {
+		return messages.ErrMissingOrEmptyInstance
+	}
+
+	if isCreate {
+		// Limit the length of the instance ID to avoid potential conflicts with state stores that have restrictive key limits.
+		const maxInstanceIDLength = 64
+		if len(instanceID) > maxInstanceIDLength {
+			return messages.ErrInstanceIDTooLong.WithFormat(maxInstanceIDLength)
+		}
+
+		// Check to see if the instance ID contains invalid characters. Valid characters are letters, digits, dashes, and underscores.
+		// See https://github.com/dapr/dapr/issues/6156 for more context on why we check this.
+		for _, c := range instanceID {
+			if !unicode.IsLetter(c) && c != '_' && c != '-' && !unicode.IsDigit(c) {
+				return messages.ErrInvalidInstanceID.WithFormat(instanceID)
+			}
+		}
+	}
+	return nil
+}
+
+func (a *UniversalAPI) getComponent(componentName string) (workflows.Workflow, error) {
```

Comment:
```suggestion
func (a *UniversalAPI) getWorkflowComponent(componentName string) (workflows.Workflow, error) {
```
There can be other methods with similar names later on for other components.

Reply:
Good call. Done.